### PR TITLE
[llvm-readobj,ELF] Support --decompress/-z

### LINKED ISSUE #82507: llvm-readelf does not support compressed sections
GNU binutils readelf and ELF Tool Chain readelf support `-z` to handle compressed sections (typically debug sections). From ELF Tool Chain's manpage:

```
     -z | --decompress
             Decompress contents of sections specified by -x or -p before
             displaying.  If the specified section is not compressed, it is
             displayed as is.
```

We are planning to migrate to LLVM's binutils-like tools and this is one issue identified during that process.
(FreeBSD bug reference https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=258872)

### PR DESCRIPTION
**Pull request by MaskRay**
When a section has the SHF_COMPRESSED flag, -p/-x dump the compressed
content by default. In GNU readelf, if --decompress/-z is specified,
-p/-x will dump the decompressed content. This patch implements the
option.

Close #82507

### CODE REVIEW

File: llvm/test/tools/llvm-readobj/ELF/decompress.test [hunk 1]
```diff
@@ -0,0 +1,67 @@
+# REQUIRES: zlib
+## Test --decompress/-z
```
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/decompress.test:2]:
```suggestion
## Test --decompress/-z.
```
Nit.

File: llvm/test/tools/llvm-readobj/ELF/decompress.test [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# REQUIRES: zlib
```
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/decompress.test:1]: Perhaps worth naming decompress-zlib.test, to allow for a zstd version too? (Same applies for the unsupported case).

**Review by jh7370 — APPROVED** [review]
One suggestion regarding zstd testing. Otherwiess LGTM.

File: llvm/test/tools/llvm-readobj/ELF/decompress.test (hunk 2 above)
Comment (MaskRay) [reply llvm/test/tools/llvm-readobj/ELF/decompress.test]: Sounds good. Done. Added a zstd version.

There is `decompress-zlib-unsupported.test` but no `decompress-zstd-unsupported.test`. The latter probably doesn't add much coverage.

Comment (jh7370) [reply llvm/test/tools/llvm-readobj/ELF/decompress.test]: That's fine, but one thought that did occur to me is that I suspect zlib is available on more machines than zstd, therefore by having a zstd unsupported test case, you'd be more likely to have coverage of both the supported and unsupported code paths on the same machine.

I'm okay if you still want to do it though.

Comment (MaskRay) [reply llvm/test/tools/llvm-readobj/ELF/decompress.test]: I think Windows is a popular OS that often misses zlib support. However, "often" may degrade to "sometimes". Anyway, I just added `decompress-zstd-unsupported.test`...

**Review by jh7370 — APPROVED** [review]
LGTM again :)

**Review by emaste — APPROVED** [review]
Thanks!

For an unsupported case I thought you might have a non-zlib/zstd case, 
```
...
    Flags: [SHF_COMPRESSED]
    Content: 03.....
```

**Comment by MaskRay** [comment]
> Thanks!
> 
> For an unsupported case I thought you might have a non-zlib/zstd case,
> 
> ```
> ...
>     Flags: [SHF_COMPRESSED]
>     Content: 03.....
> ```

Thanks. Added

**Pull request merged**